paasta_tools/cli/cmds/mesh_status.py

```diff
@@ -0,0 +1,172 @@
+
+
+def paasta_mesh_status_on_api_endpoint(
+    cluster: str,
+    service: str,
+    instance: str,
+    verbose: int,
+    system_paasta_config: SystemPaastaConfig,
+) -> Tuple[int, List[str]]:
+    client = get_paasta_oapi_client(cluster, system_paasta_config)
+    if not client:
+        print("Cannot get a paasta-api client")
+        exit(1)
+
+    try:
+        mesh_status = client.service.mesh_instance(
+            service=service,
+            instance=instance,
+            verbose=verbose,
+            include_smartstack=False,
+        )
+    except client.api_error as exc:
+        # 405 (method not allowed) is returned for instances that are not configured
+        # for the mesh, or for which getting mesh status is not supported
+        return (
+            exc.status,
+            [PaastaColors.red(exc.body if exc.status == 405 else exc.reason)],
+        )
+    except (client.connection_error, client.timeout_error) as exc:
+        return (
+            1,
+            [PaastaColors.red(f"Could not connect to API: {exc.__class__.__name__}")],
+        )
+    except Exception as e:
+        output = [PaastaColors.red(f"Exception when talking to the API:")]
+        output.extend(str(e).split("\n"))
+        return 1, output
+
+    output = []
+    if mesh_status.smartstack is not None:
+        smartstack_status_human = get_smartstack_status_human(
+            mesh_status.smartstack.registration,
+            mesh_status.smartstack.expected_backends_per_location,
+            mesh_status.smartstack.locations,
+        )
+        output.extend(smartstack_status_human)
+    if mesh_status.envoy is not None:
+        envoy_status_human = get_envoy_status_human(
+            mesh_status.envoy.registration,
+            mesh_status.envoy.expected_backends_per_location,
+            mesh_status.envoy.locations,
+        )
+        output.extend(envoy_status_human)
+
+    return 0, output
+
+
+def report_mesh_status_for_cluster(
+    cluster: str,
+    service: str,
+    instances: List[str],
+    system_paasta_config: SystemPaastaConfig,
+    verbose: int = 0,
+) -> Tuple[int, Sequence[str]]:
+    output = [f"service: {service}", f"cluster: {cluster}"]
+    return_codes = []
+
+    for instance in instances:
+        return_code, instance_output = paasta_mesh_status_on_api_endpoint(
+            cluster=cluster,
+            service=service,
+            instance=instance,
+            verbose=verbose,
+            system_paasta_config=system_paasta_config,
+        )
+
+        return_codes.append(return_code)
+        output.append(f"  instance: {PaastaColors.cyan(instance)}")
+        output.extend(["    " + line for line in instance_output])
+
+    return 1 if any(return_codes) else 0, output
+
+
+def paasta_mesh_status(args) -> int:
+    system_paasta_config = load_system_paasta_config()
+
+    return_codes = [0]
+    tasks = []
+    clusters_services_instances = apply_args_filters(args)
```

Comment:
Do you think it's worth the extra code to support multiple clusters and instances? Or could we require a cluster and instance, so we only need `paasta_mesh_status_on_api_endpoint`?

Imagining how this might be used, for debugging, I think users will likely only need a single service/instance.

Reply:
In this case I think it's fine since the code to support multi-cluster/instance output is less than the code required to validate the service/instance/cluster. I guess it depends on the principle. I don't mind making the switch if you want it.

Reply:
OK, I just noticed we were duplicating a bunch of code from the status command to do this, and was wondering if we really needed to.

For validating a service/instance/cluster, I thought we had done something similar recently with paasta logs? Might be worth making that into another util function.

Anyway, I don't feel strongly about it - up to you about what you think would be easiest to maintain.

Reply:
I changed it to be more like paasta logs.